### PR TITLE
[Environment] Meltingpot

### PR DESCRIPTION
**Pull request by matteobettini**
Wrapper for the https://github.com/google-deepmind/meltingpot multiagent library!

If you try this make sure to use the github main of tensordict as well

cc @ezhang7423 , @DJayalath

### CODE REVIEW

**Comment by matteobettini** [comment]
We should be good for review!

**Comment by ezhang7423** [comment]
Worked for me!

**Comment by ezhang7423** [comment]
I think we want to add `MeltingpotEnv` and `MeltingpotWrapper` to `./torchrl/envs/__init__.py` as well?

File: .github/workflows/test-linux-libs.yml
Comment (vmoens) [inline .github/workflows/test-linux-libs.yml:566]: IIRC this file is sorted alphabetically

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:57]: Maybe add a sentence about Meltingpot?

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:364]: why not `td.set(group, agent_tds)` which will be faster (no intermediate conversion to a tensordict)

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:334]:
```suggestion
            "It is currently unclear how to set a seed in Meltingpot"
```

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:335]:
```suggestion
            "see https://github.com/google-deepmind/meltingpot/issues/129 to track the issue."
```

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:457]: the torchrl way of doing this is to put the rendered images in a "pixels" entry in the output tensordict and a `from_pixels=True` arg in constructor. So far, no env supports render by design.

**Review by vmoens — COMMENTED** [review]
Thanks for this, superb!

File: torchrl/envs/libs/meltingpot.py
Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py:457]: Meltingpot has rgb images by default part of the observation dictionary under the keys : "WORLD.RGB" and "RGB".

This is just a util to show those to the GUI or save them to file.

I thought it would be helpful since MeltingPot itself does not have it and i need to call `env.render()` in BenchMARL, but if you think it is not helpful i'll move it to BenchMARL.

The torchrl way seems to be related to input output. This function is not and can be called at any time

Comment (vmoens) [reply torchrl/envs/libs/meltingpot.py:457]: Oh I get it you're not rendering anything but just using matplotlib to display it.

I don't think that should be within this env, because that makes the whole class dependent on matplotlib and that doesn't seem super neat (one more extra dep to handle etc).

Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py:457]: matplotlib is imported only if you call the function.

Which seems fair to me no?

Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py:457]: actually i can make it so that matplotlib is imported only is the users calls the function and asks for GUI

Comment (vmoens) [reply torchrl/envs/libs/meltingpot.py:457]: Sure, but we need to test it if we integrate it no?
Plus that seems fairly simple and orthogonal to what the env is doing (not required to run it).
To me that shouldn't be part of the env wrapper

Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py:457]: I think that prividing the method done right is a plus for users so I do not really agree.

Anyway I'll just put it in BenchMARL

Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py:457]: what do you think about a method called `get_rgb_image`, which returns an image of the env at any time?

File: torchrl/envs/libs/meltingpot.py
Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py]: Thanks for catching this!

**Comment by matteobettini** [comment]
In the last commit I modified and tested the `render` function (changed to `get_rgb_image`)

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [reply torchrl/envs/libs/meltingpot.py]: yeah sure that sounds reasonable to me!
the thing I was skeptical about was the matplotlib dependency.

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:447]: Being torchrl I would expect this to be a tensor

Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py:447]: The reasoning is to avoid the conversion cost since the most common use case i envision is rendering

Comment (vmoens) [reply torchrl/envs/libs/meltingpot.py:447]: We only use tensor in the lib. torch.as_tensor has a negligible overhead and I think matplotlib will work fine with a tensor too.

File: .github/unittest/linux_libs/scripts_meltingpot/environment.yml
Comment (vmoens) [inline .github/unittest/linux_libs/scripts_meltingpot/environment.yml:20]: shouldn't this be dm-meltingpot?

Comment (matteobettini) [reply .github/unittest/linux_libs/scripts_meltingpot/environment.yml:20]: oh wow you are right, there is a shadow PyPi package

File: test/test_libs.py
Comment (vmoens) [inline test/test_libs.py:3496]: Shoudn't we use the WORLD_PREFIX variable here?

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:56]: Shouldn't we use the WORLD_PREFIX here?

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:58]: Shouldn't we use the WORLD_PREFIX here?

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:136]: ditto

File: torchrl/envs/libs/meltingpot.py
Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py:118]: This is the meltingpot format though, they also use dictionaries as the input/output, changing names would be surprising to users

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:452]: Shouldn't we use the WORLD_PREFIX here?

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:507]: ditto

File: torchrl/envs/libs/meltingpot.py
Comment (vmoens) [inline torchrl/envs/libs/meltingpot.py:525]: ditto

File: torchrl/envs/libs/meltingpot.py
Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py:118]: What kind of issues? What can we do to prevent them?

Comment (matteobettini) [reply torchrl/envs/libs/meltingpot.py:118]: What I could also do is remove the WORLD_PREFIX.

Sine it is already not part of any agent group, it is implied that it is shared

Comment (vmoens) [reply torchrl/envs/libs/meltingpot.py:118]: by default flatten uses the "." separator so a flatten/unflatten sequence will nest the WOLD.SMTH in another tensordict.
Upon reflection I think serialization will be fine.

**Comment by matteobettini** [comment]
I am not getting what is happening here in the CI

`AttributeError: module 'meltingpot' has no attribute 'substrate'`

It seems like meltingpot has been installed correctly...but yet this happens

File: .github/workflows/test-linux-libs.yml
Comment (vmoens) [inline .github/workflows/test-linux-libs.yml:276]: you should have python 3.11

**Comment by matteobettini** [comment]
Now pip is complaining about a conflict which I do not see since the only dep i am installing is dm_meltinglot.

it is kind of difficult for me to fix these CI issues as every time i require the ci approval to try something out and i cannot reproduce locally ofc.

 I checked the tests locally and they pass so it is just a matter to convince the CI to install the lib, if you could help me with that

**Comment by matteobettini** [comment]
Now it seems that meltingpot (which has python 3.11,3.12 listed as compatible on the github) requires dmlab which is not available for python >= 3.11

File: .github/unittest/linux_libs/scripts_meltingpot/install.sh
Comment (vmoens) [inline .github/unittest/linux_libs/scripts_meltingpot/install.sh:62]:
```suggestion
pip3 install dmlab2d
```

**Comment by ezhang7423** [comment]
Anything I can do to help @vmoens?

**Comment by vmoens** [comment]
Still blocked by dmlab2d, not sure why but it can't find a suitable package...

**Comment by matteobettini** [comment]
Let me trying going back to python 3.9, since with that we were installing all packages

**Comment by matteobettini** [comment]
I think the ci is generally crashing now so I cannot tell if it worked, but at least it seems to install

**Comment by matteobettini** [comment]
The latest tag of meltingpot won't work with python 3.9, we need to let pip figure it out and download the latest compatible

File: .github/unittest/linux_libs/scripts_meltingpot/install.sh
Comment (vmoens) [inline .github/unittest/linux_libs/scripts_meltingpot/install.sh:64]:
```suggestion
pip3 install git+https://github.com/google-deepmind/meltingpot@${LATEST_TAG}
```

File: .github/unittest/linux_libs/scripts_meltingpot/install.sh
Comment (vmoens) [inline .github/unittest/linux_libs/scripts_meltingpot/install.sh:64]:
```suggestion
pip3 install git+https://github.com/google-deepmind/meltingpot@${LATEST_TAG}
```

File: .github/workflows/test-linux-libs.yml
Comment (vmoens) [inline .github/workflows/test-linux-libs.yml:265]:
```suggestion
      gpu-arch-version: "12.1"
```

File: .github/unittest/linux_libs/scripts_meltingpot/setup_env.sh
Comment (vmoens) [inline .github/unittest/linux_libs/scripts_meltingpot/setup_env.sh:19]:
```suggestion
yum update -y
yum install glibc.x86_64 -y
```

File: .github/unittest/linux_libs/scripts_meltingpot/setup_env.sh
Comment (vmoens) [inline .github/unittest/linux_libs/scripts_meltingpot/setup_env.sh:8]:
```suggestion
set -e -v
```

File: .github/workflows/test-linux-libs.yml
Comment (vmoens) [inline .github/workflows/test-linux-libs.yml:266]:
```suggestion
      docker-image: cuda:12.4.1-runtime-ubuntu22.04
```

File: .github/unittest/linux_libs/scripts_meltingpot/setup_env.sh
Comment (vmoens) [inline .github/unittest/linux_libs/scripts_meltingpot/setup_env.sh:19]:
```suggestion
```

**Comment by matteobettini** [comment]
What should we do with this? Is there any way you can allow me to run the CI?

**Comment by vmoens** [comment]
It's a docker image issue. Leave it with me!

File: .github/unittest/linux_libs/scripts_meltingpot/setup_env.sh
Comment (vmoens) [inline .github/unittest/linux_libs/scripts_meltingpot/setup_env.sh:11]:
```suggestion
apt-get install -y git wget g++ gcc libglfw3 libgl1-mesa-glx libosmesa6 libglew-dev libglvnd0 libgl1 libglx0 libegl1 libgles2 curl
```

**Review by vmoens — APPROVED** [review]
LGTM

**Comment by ezhang7423** [comment]
Congrats @matteobettini @vmoens !!

**Pull request merged**
